src/services/resource.js

```diff
@@ -0,0 +1,1297 @@
+          removed: resources.removed,
+          reduced: [
+            ...resources.reduced,
+            {
+              genericId: resource.genericId,
+              resourceId: resource.resourceId,
+              sourceFields: resource.sourceFields
+                .filter((sourceField) => matching.sourceFields.includes(sourceField)),
+            },
+          ],
+        };
+      }
+
+      return resources;
+    }, { removed: [], reduced: [] });
+
+  // collect the intersection of the expanded and reduced datasets to generate the delta dataset.
+  const deltaFromExpanded = (newExpandedResources.expanded || [])
+    .filter((neResource) => (removedReducedResources.reduced || [])
+      .filter((rrResource) => neResource.genericId === rrResource.genericId
+        && neResource.resourceId === rrResource.resourceId)
+      .length > 0);
+  const deltaFromReduced = (removedReducedResources.reduced || [])
+    .filter((rrResource) => (newExpandedResources.expanded || [])
+      .filter((neResource) => neResource.genericId === rrResource.genericId
+        && neResource.resourceId === rrResource.resourceId)
+      .length > 0);
+
+  const resourceActions = {};
+  // Generate the delta dataset
+  resourceActions.delta = deltaFromExpanded
+    ?.reduce((delta, resource) => {
+      const exists = delta.find((r) => r.genericId === resource.genericId
+        && r.resourceId === resource.resourceId);
+      const fromReduced = deltaFromReduced
+        ?.find((r) => r.genericId === resource.genericId
+        && r.resourceId === resource.resourceId);
+      const fromOriginal = currentResources
+        .find((r) => r.genericId === resource.genericId
+        && r.resourceId === resource.resourceId);
+      const deltaSourceFields = resource.sourceFields
+        .filter((sourceField) => !fromOriginal.sourceFields.includes(sourceField)
+          || fromReduced?.sourceFields.includes(sourceField));
+      if (exists) {
+        exists.sourceFields = deltaSourceFields;
+        return delta;
+      }
+      return [
+        ...delta,
+        {
+          ...resource,
+          sourceFields: deltaSourceFields,
+        },
+      ];
+    }, [])
+    ?.map((resource) => ({
+      ...resource,
+    }));
+  // Remove the records of the delta dataset from the expanded dataset.
+  resourceActions.expanded = newExpandedResources.expanded
+    ?.filter((neResource) => resourceActions.delta
+      .filter((dResource) => dResource.genericId === neResource.genericId
+        && dResource.resourceId === neResource.resourceId)
+      .length === 0)
+    .map((resource) => ({
+      ...resource,
+    }));
+  // Remove the records of the delta dataset from the reduced dataset.
+  resourceActions.reduced = removedReducedResources.reduced
+    .filter((rrResource) => (resourceActions.delta || [])
+      .filter((dResource) => dResource.genericId === rrResource.genericId
+        && dResource.resourceId === rrResource.resourceId)
+      .length === 0);
```

Comment:
here as well, replacing `filter`.`length` with `some`:

```suggestion
resourceActions.reduced = removedReducedResources.reduced?.filter((rrResource) => (
  !resourceActions.delta?.some((dResource) => (
    dResource.genericId === rrResource.genericId && dResource.resourceId === rrResource.resourceId
  ))
));
```